yarn: resource manager should make sure spnego keytab exists

I'm working on deploying a TDP cluster where HDFS is already installed on an external bay. When running the [/playbooks/meta/yarn.yml](https://github.com/TOSIT-IO/tdp-collection/blob/87b5e78bf550de6e87874b48896e6b621fbb4851/playbooks/meta/yarn.yml) playbook, at first, everything seems to be working fine but the resource manager services will fail after a couple seconds with the following error found in `/var/log/hadoop/yarn/hadoop-yarn-resourcemanager-[hostname].log` :

```
2022-08-18 10:18:16,891 FATAL org.apache.hadoop.yarn.server.resourcemanager.ResourceManager: Error starting ResourceManager
org.apache.hadoop.yarn.webapp.WebAppException: Error starting http server
        at org.apache.hadoop.yarn.webapp.WebApps$Builder.start(WebApps.java:443)
        at org.apache.hadoop.yarn.server.resourcemanager.ResourceManager.startWepApp(ResourceManager.java:1189)
        at org.apache.hadoop.yarn.server.resourcemanager.ResourceManager.serviceStart(ResourceManager.java:1298)
        at org.apache.hadoop.service.AbstractService.start(AbstractService.java:194)
        at org.apache.hadoop.yarn.server.resourcemanager.ResourceManager.main(ResourceManager.java:1493)
Caused by: java.io.IOException: Problem starting http server
        at org.apache.hadoop.http.HttpServer2.start(HttpServer2.java:1165)
        at org.apache.hadoop.yarn.webapp.WebApps$Builder.start(WebApps.java:439)
        ... 4 more
Caused by: javax.servlet.ServletException: javax.servlet.ServletException: Keytab does not exist: /etc/security/keytabs/spnego.service.keytab
        at org.apache.hadoop.security.authentication.server.KerberosAuthenticationHandler.init(KerberosAuthenticationHandler.java:179)
        at org.apache.hadoop.security.token.delegation.web.DelegationTokenAuthenticationHandler.init(DelegationTokenAuthenticationHandler.java:130)
        at org.apache.hadoop.security.authentication.server.AuthenticationFilter.initializeAuthHandler(AuthenticationFilter.java:194)
        at org.apache.hadoop.security.token.delegation.web.DelegationTokenAuthenticationFilter.initializeAuthHandler(DelegationTokenAuthenticationFilter.java:214)
        at org.apache.hadoop.security.authentication.server.AuthenticationFilter.init(AuthenticationFilter.java:180)
        at org.apache.hadoop.security.token.delegation.web.DelegationTokenAuthenticationFilter.init(DelegationTokenAuthenticationFilter.java:180)
        at org.apache.hadoop.yarn.server.security.http.RMAuthenticationFilter.init(RMAuthenticationFilter.java:53)
        at org.eclipse.jetty.servlet.FilterHolder.initialize(FilterHolder.java:139)
        at org.eclipse.jetty.servlet.ServletHandler.initialize(ServletHandler.java:873)
        at org.eclipse.jetty.servlet.ServletContextHandler.startContext(ServletContextHandler.java:349)
        at org.eclipse.jetty.server.handler.ContextHandler.doStart(ContextHandler.java:778)
        at org.eclipse.jetty.servlet.ServletContextHandler.doStart(ServletContextHandler.java:262)
        at org.eclipse.jetty.util.component.AbstractLifeCycle.start(AbstractLifeCycle.java:68)
        at org.eclipse.jetty.util.component.ContainerLifeCycle.start(ContainerLifeCycle.java:131)
        at org.eclipse.jetty.util.component.ContainerLifeCycle.doStart(ContainerLifeCycle.java:113)
        at org.eclipse.jetty.server.handler.AbstractHandler.doStart(AbstractHandler.java:61)
        at org.eclipse.jetty.server.handler.ContextHandlerCollection.doStart(ContextHandlerCollection.java:161)
        at org.eclipse.jetty.util.component.AbstractLifeCycle.start(AbstractLifeCycle.java:68)
        at org.eclipse.jetty.util.component.ContainerLifeCycle.start(ContainerLifeCycle.java:131)
        at org.eclipse.jetty.util.component.ContainerLifeCycle.doStart(ContainerLifeCycle.java:113)
        at org.eclipse.jetty.server.handler.AbstractHandler.doStart(AbstractHandler.java:61)
        at org.eclipse.jetty.util.component.AbstractLifeCycle.start(AbstractLifeCycle.java:68)
        at org.eclipse.jetty.util.component.ContainerLifeCycle.start(ContainerLifeCycle.java:131)
        at org.eclipse.jetty.server.Server.start(Server.java:422)
        at org.eclipse.jetty.util.component.ContainerLifeCycle.doStart(ContainerLifeCycle.java:105)
        at org.eclipse.jetty.server.handler.AbstractHandler.doStart(AbstractHandler.java:61)
        at org.eclipse.jetty.server.Server.doStart(Server.java:389)
        at org.eclipse.jetty.util.component.AbstractLifeCycle.start(AbstractLifeCycle.java:68)
        at org.apache.hadoop.http.HttpServer2.start(HttpServer2.java:1134)
        ... 5 more
Caused by: javax.servlet.ServletException: Keytab does not exist: /etc/security/keytabs/spnego.service.keytab
        at org.apache.hadoop.security.authentication.server.KerberosAuthenticationHandler.init(KerberosAuthenticationHandler.java:139)
        ... 33 more
```

I fixed the error on my side by making sure the spnego file is created on each node of the yarn_rm group.

I did not meet this error when deploying Yarn after having deployed HDFS with TDP on the same cluster because the spnego.service.keytab file is created on the hdfs_nn group of hosts, which happens to contain the same hosts as the yarn_rm group of hosts. 

If someone decides to separate hdfs_nn and yarn_nn on different nodes, yarn resource manager will fail.

👉 The role [/roles/yarn/resourcemanager/tasks/kerberos.yml](https://github.com/TOSIT-IO/tdp-collection/blob/87b5e78bf550de6e87874b48896e6b621fbb4851/roles/yarn/resourcemanager/tasks/kerberos.yml) should create the spnego keytab file like [/roles/yarn/nodemanager/tasks/kerberos.yml](https://github.com/TOSIT-IO/tdp-collection/blob/87b5e78bf550de6e87874b48896e6b621fbb4851/roles/yarn/nodemanager/tasks/kerberos.yml) does.

--- a/roles/yarn/resourcemanager/tasks/kerberos.yml
+++ b/roles/yarn/resourcemanager/tasks/kerberos.yml
@@ -10,24 +10,46 @@
     name: tosit.tdp.hadoop.common
     tasks_from: kerberos
 
-- import_role:
-    name: tosit.tdp.utils.kerberos
-    tasks_from: create_principal_keytab
-  vars:
-    principal: "rm/{{ ansible_fqdn }}"
-    keytab: "rm.service.keytab"
-    user: "{{ yarn_user }}"
-    group: "{{ hadoop_group }}"
-    mode: "600"
+- block:
+    - import_role:
+        name: tosit.tdp.utils.kerberos
+        tasks_from: create_principal_keytab
+      vars:
+        principal: "rm/{{ ansible_fqdn }}"
+        keytab: "rm.service.keytab"
+        user: "{{ yarn_user }}"
+        group: "{{ hadoop_group }}"
+        mode: "600"
+
+    - import_role:
+        name: tosit.tdp.utils.kerberos
+        tasks_from: create_principal_keytab
+      vars:
+        principal: "HTTP/{{ ansible_fqdn }}"
+        keytab: "spnego.service.keytab"
+        user: "root"
+        group: "{{ hadoop_group }}"
+        mode: "640"
   when: krb_create_principals_keytabs
 
-- import_role:
-    name: tosit.tdp.utils.kerberos
-    tasks_from: check_secure_keytab
-  vars:
-    principal: "rm/{{ ansible_fqdn }}"
-    keytab: "rm.service.keytab"
-    user: "{{ yarn_user }}"
-    group: "{{ hadoop_group }}"
-    mode: "600"
+- block:
+    - import_role:
+        name: tosit.tdp.utils.kerberos
+        tasks_from: check_secure_keytab
+      vars:
+        principal: "rm/{{ ansible_fqdn }}"
+        keytab: "rm.service.keytab"
+        user: "{{ yarn_user }}"
+        group: "{{ hadoop_group }}"
+        mode: "600"
+
+    - import_role:
+        name: tosit.tdp.utils.kerberos
+        tasks_from: check_secure_keytab
+      vars:
+        principal: "HTTP/{{ ansible_fqdn }}"
+        keytab: "spnego.service.keytab"
+        user: "root"
+        group: "{{ hadoop_group }}"
+        mode: "640"
   when: not krb_create_principals_keytabs

